CBMC does not warn when indexing off a void*

<!---
Thank you for reporting a problem and suggesting improvements. Please provide the below information to make sure we can effectively deal with the issue reported. For the most precise version information, see the first line of console output or run with --version. Please attach or include example code that allows us to reproduce the problem.
--->

```
#include <assert.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include <stdbool.h>

int main(int argc, char* argv[]) {
    void* p = malloc(2);
    void* q = &p[1];
    printf("%p\n", p);
    printf("%p\n", q);
    return 0;
}
```

CBMC version: `5.12 (cbmc-5.12-104-g97993174d)`
Operating system: OSX
Exact command line resulting in the issue:
```
cbmc  --flush --bounds-check --div-by-zero-check --float-overflow-check --nan-check --pointer-check --pointer-overflow-check --signed-overflow-check --undefined-shift-check --unsigned-overflow-check --unwind 1 --unwinding-assertions test.c
```
What behaviour did you expect: I expected it to say that you can't index using a `void*`, since void has no size
What happened instead: 
```
** 0 of 49 failed (1 iterations)
VERIFICATION SUCCESSFUL
```

A couple of experiments with different compilers:
- Visual Studio does not accept this program ("error C2036: 'void *': unknown size")
- GCC warns ("warning: dereferencing 'void *' pointer"), but compiles this to pointer arithmetic (adding 1)
- Clang doesn't even warn, compiles this to pointer arithmetic (adding 1)

GCC and Clang's behaviour is consistent with GCC's documentation: https://gcc.gnu.org/onlinedocs/gcc/Pointer-Arith.html.

CBMC also turns this into pointer arithmetic:
```
p!0@1#2 == (const void *)dynamic_object1
q!0@1#2 == (const void *)dynamic_object1 + 1l
```

In several places in the back-end we have comments that highlight that we are treating this the way GCC's documentation describes. Maybe we should warn about this in the front-end when not in GCC mode?

--- a/regression/ansi-c/sizeof5/test.desc
+++ b/regression/ansi-c/sizeof5/test.desc
@@ -1,4 +1,4 @@
-CORE
+CORE gcc-only
 main.c
 
 ^EXIT=0$

--- a/regression/cbmc-cpp/Function_Arguments1/test.desc
+++ b/regression/cbmc-cpp/Function_Arguments1/test.desc
@@ -1,4 +1,4 @@
-CORE
+CORE gcc-only
 main.cpp
 
 ^EXIT=0$

--- a/regression/cbmc/null6/main.c
+++ b/regression/cbmc/null6/main.c
@@ -1,4 +1,4 @@
-void *guard_malloc_counter = 0;
+char *guard_malloc_counter = 0;
 
 void *my_malloc(int size)
 {

--- a/regression/cbmc/ptr_arithmetic_on_null/test.desc
+++ b/regression/cbmc/ptr_arithmetic_on_null/test.desc
@@ -1,4 +1,4 @@
-CORE
+CORE gcc-only
 main.c
 
 ^\[main.assertion.1\] line .* assertion \(void \*\)0 != \(void \*\)0 \+ \(.*\)1: SUCCESS$

--- a/regression/cbmc/void_pointer1/test.desc
+++ b/regression/cbmc/void_pointer1/test.desc
@@ -1,4 +1,4 @@
-CORE
+CORE gcc-only
 main.c
 --pointer-check
 ^EXIT=0$

--- a/regression/cbmc/void_pointer2/test.desc
+++ b/regression/cbmc/void_pointer2/test.desc
@@ -1,4 +1,4 @@
-CORE
+CORE gcc-only
 main.c
 --pointer-check --no-simplify --unwind 3
 ^EXIT=0$

--- a/regression/cbmc/void_pointer3/test.desc
+++ b/regression/cbmc/void_pointer3/test.desc
@@ -1,4 +1,4 @@
-CORE broken-smt-backend
+CORE broken-smt-backend gcc-only
 main.c
 
 ^EXIT=0$

--- a/regression/cbmc/void_pointer5/test.desc
+++ b/regression/cbmc/void_pointer5/test.desc
@@ -1,4 +1,4 @@
-CORE broken-smt-backend
+CORE broken-smt-backend gcc-only
 main.c
 
 ^EXIT=10$

--- a/regression/cbmc/void_pointer6/test.desc
+++ b/regression/cbmc/void_pointer6/test.desc
@@ -1,4 +1,4 @@
-CORE
+CORE gcc-only
 main.c
 --pointer-check
 ^EXIT=10$

--- a/regression/cbmc/void_pointer7/test.desc
+++ b/regression/cbmc/void_pointer7/test.desc
@@ -1,4 +1,4 @@
-CORE
+CORE gcc-only
 main.c
 --pointer-check
 ^EXIT=0$

--- a/regression/contracts/assigns_enforce_conditional_void_target/test.desc
+++ b/regression/contracts/assigns_enforce_conditional_void_target/test.desc
@@ -1,7 +1,7 @@
 CORE
 main.c
 --enforce-contract foo
-^.* error: void-typed expressions not allowed as assigns clause targets$
+^.* error: (void-typed expressions not allowed as assigns clause targets|dereferencing void pointer)$
 ^CONVERSION ERROR$
 ^EXIT=(1|64)$
 ^SIGNAL=0$

--- a/regression/contracts/assigns_enforce_conditional_void_target_list/test.desc
+++ b/regression/contracts/assigns_enforce_conditional_void_target_list/test.desc
@@ -1,7 +1,7 @@
 CORE
 main.c
 --enforce-contract foo
-^.* error: void-typed expressions not allowed as assigns clause targets$
+^.* error: (void-typed expressions not allowed as assigns clause targets|dereferencing void pointer)$
 ^CONVERSION ERROR$
 ^EXIT=(1|64)$
 ^SIGNAL=0$

--- a/regression/contracts/assigns_enforce_side_effects_1/test.desc
+++ b/regression/contracts/assigns_enforce_side_effects_1/test.desc
@@ -1,9 +1,8 @@
 CORE
 main.c
 --enforce-contract foo
-^.*error: void-typed expressions not allowed as assigns clause targets$
-^.*error: side-effects not allowed in assigns clause targets$
-^.*error: ternary expressions not allowed in assigns clause targets$
+activate-multi-line-match
+.*error: (dereferencing void pointer|void-typed expressions not allowed as assigns clause targets\n.*\n.*error: side-effects not allowed in assigns clause targets\n.*\n.*error: ternary expressions not allowed in assigns clause targets\n)
 ^CONVERSION ERROR$
 ^EXIT=(1|64)$
 ^SIGNAL=0$

--- a/regression/goto-cl/void-pointer/main.c
+++ b/regression/goto-cl/void-pointer/main.c
@@ -1,0 +1,15 @@
+void *malloc(__CPROVER_size_t);
+int printf(const char *, ...);
+
+int main(int argc, char *argv[])
+{
+  void *p = malloc(2);
+  printf("%p\n", p);
+#ifdef VOID1
+  (void)*p;
+#else
+  void *q = &p[1];
+  printf("%p\n", q);
+#endif
+  return 0;
+}

--- a/regression/goto-cl/void-pointer/ptr-arithmetic.desc
+++ b/regression/goto-cl/void-pointer/ptr-arithmetic.desc
@@ -1,0 +1,9 @@
+CORE
+main.c
+
+error: pointer arithmetic with unknown object size
+^CONVERSION ERROR$
+^EXIT=(64|1)$
+^SIGNAL=0$
+--
+Invariant check failed

--- a/regression/goto-cl/void-pointer/test.desc
+++ b/regression/goto-cl/void-pointer/test.desc
@@ -1,0 +1,9 @@
+CORE
+main.c
+-DVOID1
+error: dereferencing void pointer
+^CONVERSION ERROR$
+^EXIT=(64|1)$
+^SIGNAL=0$
+--
+Invariant check failed

--- a/src/ansi-c/c_typecheck_expr.cpp
+++ b/src/ansi-c/c_typecheck_expr.cpp
@@ -1784,6 +1784,15 @@ void c_typecheck_baset::typecheck_expr_dereference(exprt &expr)
   else if(op_type.id()==ID_pointer)
   {
     expr.type() = to_pointer_type(op_type).base_type();
+
+    if(
+      expr.type().id() == ID_empty &&
+      config.ansi_c.mode == configt::ansi_ct::flavourt::VISUAL_STUDIO)
+    {
+      error().source_location = expr.source_location();
+      error() << "dereferencing void pointer" << eom;
+      throw 0;
+    }
   }
   else
   {
@@ -3713,6 +3722,14 @@ void c_typecheck_baset::typecheck_arithmetic_pointer(const exprt &expr)
   else if(
     base_type.id() == ID_union_tag &&
     follow_tag(to_union_tag_type(base_type)).is_incomplete())
+  {
+    error().source_location = expr.source_location();
+    error() << "pointer arithmetic with unknown object size" << eom;
+    throw 0;
+  }
+  else if(
+    base_type.id() == ID_empty &&
+    config.ansi_c.mode == configt::ansi_ct::flavourt::VISUAL_STUDIO)
   {
     error().source_location = expr.source_location();
     error() << "pointer arithmetic with unknown object size" << eom;

